homeassistant/components/monzo/__init__.py

```diff
@@ -0,0 +1,210 @@
+    hass.data[DOMAIN][DATA_MONZO_LISTENER][
+        config_entry.entry_id] = async_track_time_interval(
+            hass, refresh_sensors,
+            hass.data.get(DATA_MONZO_CONFIG, {}).get(CONF_SCAN_INTERVAL,
+                                                     DEFAULT_SCAN_INTERVAL))
+
+    _LOGGER.debug("async_setup_monzo is done")
+
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
+    """Unload a Monzo config entry."""
+    for component in ('sensor',):
+        await hass.config_entries.async_forward_entry_unload(
+            config_entry, component)
+
+    hass.data[DOMAIN][DATA_MONZO_CLIENT].pop(config_entry.entry_id)
+
+    remove_listener = hass.data[DOMAIN][DATA_MONZO_LISTENER].pop(
+        config_entry.entry_id)
+    remove_listener()
+
+    return True
+
+
+class MonzoObject:
+    """Define a generic Monzo object."""
+
+    def __init__(self, hass, sensor_conditions, config_entry):
+        """Initialize."""
+        from monzo.monzo import Monzo
+        from monzo.auth import MonzoOAuth2Client
+
+        self.data = {}
+        self._hass = hass
+        self.sensor_conditions = sensor_conditions
+        self.config_entry = config_entry
+
+        client_id = config_entry.data['tokens'][CONF_CLIENT_ID]
+        client_secret = config_entry.data['tokens'][CONF_CLIENT_SECRET]
+        access_token = config_entry.data['tokens'][CONF_ACCESS_TOKEN]
+        refresh_token = config_entry.data['tokens'][CONF_REFRESH_TOKEN]
+        last_saved_at = config_entry.data['tokens'].get(CONF_LAST_SAVED_AT)
+        expires_at = config_entry.data['tokens'].get(CONF_EXPIRES_AT,
+                                                     last_saved_at)
+
+        oauth_client = MonzoOAuth2Client(
+            client_id=client_id,
+            client_secret=client_secret,
+            access_token=access_token,
+            refresh_token=refresh_token,
+            expires_at=expires_at,
+            refresh_callback=self.update_config_entry)
+
+        self.client = Monzo.from_oauth_session(oauth_client)
+
+    async def async_update(self):
+        """Update sensor data."""
+        if any(sensor_type in self.sensor_conditions for
+               sensor_type in (TYPE_BALANCE, TYPE_DAILY_SPEND)):
+            account = await self._hass.async_add_executor_job(
+                self.client.get_first_account)
+
+            account_id = account['id']
+            self.data[DATA_ACCOUNTID] = account_id
+
+            balance = await self._hass.async_add_executor_job(
+                self.client.get_balance, account_id)
+
+            self.data[DATA_BALANCE] = balance
+
+        if TYPE_POTS in self.sensor_conditions:
+            all_pots = await self._hass.async_add_executor_job(
+                self.client.get_pots)
+
+            open_pots = [pot for pot in all_pots['pots'] if not pot['deleted']]
+            self.data[DATA_POTS] = open_pots
+
+    @callback
+    def update_config_entry(self, new_token):
```

Comment:
This only gets called by self.client from within one of the executor jobs in async_update so would only get called from an async context.

I think the above fix makes this kosher but shout if it's not.

Reply:
That means it will run in another thread than the event loop, ie non async context.